File: kotlin/src/main/kotlin/behavioral/chain/MainChain.kt
package behavioral.chain

internal interface Element {
    fun operation(i: Int)
}

internal class LastElement : Element {
    override fun operation(i: Int) {
        println("last $i operation")
    }
}

internal class IterElement(private val element: Element) : Element {
    override fun operation(i: Int) {
        println("$i operation")
        element.operation(i + 1)
    }
}

fun main(args: Array<String>) {
    val lastElement: Element = LastElement()
    val iterElement2: Element = IterElement(lastElement)
    val iterElement1: Element = IterElement(iterElement2)
    iterElement1.operation(1)
}

File: python/behavioral/chain/MainChain.py
from abc import ABC, abstractmethod

class Element(ABC):
    @abstractmethod
    def operation(self, i: int):
        pass

class LastElement(Element):
    def operation(self, i: int):
        print("last " + str(i) + " operation")



class IterElement(Element):
    def __init__(self, element: Element):
        self.__element = element

    def operation(self, i: int):
        print(str(i) + " operation");
        self.__element.operation(i + 1)


def main():
    lastElement = LastElement()
    iterElement2 = IterElement(lastElement)
    iterElement1 = IterElement(iterElement2)
    iterElement1.operation(1)

main()
